### PR TITLE
[NTGDI][FREETYPE] Simplify font rendering

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Simplify the font rendering algorithm.
JIRA issue: [CORE-11848](https://jira.reactos.org/browse/CORE-11848)

## Proposed changes

- Add `ftGdiGetRealGlyph` and `ftGdiGetTextWidth` helper functions.
- Optimize font rendering.

## TODO

- [x] Fix bugs.
- [x] Do small tests.
- [x] Do big tests.
- [x] Real world tests.
- [x] Speed.

### CODE REVIEW

File: win32ss/gdi/ntgdi/freetype.c [hunk 1]
```diff
@@ -6228,118 +6237,31 @@ IntExtTextOutW(
...
+                     NULL,
+                     NULL,
+                     (CLIPOBJ *)&dc->co,
+                     NULL,
+                     &DestRect,
+                     &SourcePoint,
+                     &SourcePoint,
+                     &dc->eboBackground.BrushObject,
+                     &BrushOrigin,
+                     ROP4_FROM_INDEX(R3_OPINDEX_PATCOPY));
```
Comment (katahiromz) [inline win32ss/gdi/ntgdi/freetype.c:6263]: VirtualBox_ReactOS_09_11_2022_10_54_40
Why is the background not filled correctly?

Comment (HBelusca) [reply win32ss/gdi/ntgdi/freetype.c:6263]: Was it filled correctly without the PR?

Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c:6263]: Step 1:


Step 2:


Step 3:


Step 4:


Step 5:


Refreshment is something wrong...

Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c:6263]: > Was it filled correctly without the PR?

Before this PR, the background is filled in main loop.

File: win32ss/gdi/ntgdi/freetype.c [hunk 2]
```diff
@@ -6413,39 +6334,6 @@ IntExtTextOutW(
         DPRINT("TextTop: %lu\n", TextTop);
         DPRINT("Advance: %d\n", realglyph->root.advance.x);
 
-        if ((fuOptions & ETO_OPAQUE) && !plf->lfItalic)
```
Comment (katahiromz) [inline win32ss/gdi/ntgdi/freetype.c:6416]: Here.

**Comment by katahiromz** [comment]
Underline is added for debugging.
VirtualBox_ReactOS_09_11_2022_12_31_27

**Comment by katahiromz** [comment]
Currently, `TextWidth` is inaccurate.

**Comment by katahiromz** [comment]
VirtualBox_ReactOS_09_11_2022_17_29_21

image
Underline not connected.

**Comment by katahiromz** [comment]
Blit cache problem??? --- there is a problem as seen the underline being overwritten unexpectedly.

I cannot fix that problem. Closed.

**Comment by katahiromz** [comment]
VirtualBox_ReactOS_21_11_2022_08_53_20

**Comment by katahiromz** [comment]
Test KVM BEFORE: 79b0fce https://build.reactos.org/#/builders/6/builds/32468 id=84948
Test KVM AFTER: a00e7e5 https://build.reactos.org/#/builders/6/builds/32469 id=???
https://reactos.org/testman/compare.php?ids=84948,???

Test VBox BEFORE: 79b0fce https://build.reactos.org/#/builders/3/builds/19945 id=84949
Test VBox AFTER: a00e7e5  https://build.reactos.org/#/builders/3/builds/19946 id=???
https://reactos.org/testman/compare.php?ids=84949,???

**Comment by katahiromz** [comment]
Test KVM BEFORE: 79b0fce https://build.reactos.org/#/builders/6/builds/32468 id=84948
Test KVM AFTER: d36dd6b
https://reactos.org/testman/compare.php?ids=84948,???

Test VBox BEFORE: 79b0fce https://build.reactos.org/#/builders/3/builds/19945 id=84949
Test VBox AFTER: d36dd6b
https://reactos.org/testman/compare.php?ids=84949,???

**Comment by katahiromz** [comment]
Try again.

Test KVM BEFORE: 79b0fce https://build.reactos.org/#/builders/6/builds/32468 id=84948
Test KVM AFTER: b811669 id=???
https://reactos.org/testman/compare.php?ids=84948,???

Test VBox BEFORE: 79b0fce https://build.reactos.org/#/builders/3/builds/19945 id=84949
Test VBox AFTER: b811669 id=???
https://reactos.org/testman/compare.php?ids=84949,???

File: win32ss/gdi/ntgdi/freetype.c [hunk 3]
```diff
@@ -5880,6 +5894,59 @@ ScaleLong(LONG lValue, PFLOATOBJ pef)
...
+            return FALSE;
+
+        /* Retrieve kerning distance */
+        if (use_kerning && previous && glyph_index)
+        {
+            FT_Get_Kerning(face, previous, glyph_index, 0, &delta);
+            TextLeft += delta.x;
+        }
+
+        TextLeft += realglyph->root.advance.x >> 10;
```
Comment (learn-more) [inline win32ss/gdi/ntgdi/freetype.c:5936]: Would it make sense to not round per character, but only round after the entire string is calculated?
E.g. have `TextLeft` in high precision?

File: win32ss/gdi/ntgdi/freetype.c [hunk 4]
```diff
@@ -5880,6 +5894,59 @@ ScaleLong(LONG lValue, PFLOATOBJ pef)
     return lValue;
 }
 
+/* Calculate width of the text. */
+static inline
+BOOL
+APIENTRY
+ftGdiGetTextWidth(
```
Comment (learn-more) [inline win32ss/gdi/ntgdi/freetype.c:5901]: This function seems to require the freetype lock held, please guard that with an assert :)

File: win32ss/gdi/ntgdi/freetype.c [hunk 5]
```diff
@@ -6071,7 +6123,7 @@ IntExtTextOutW(
     }
 
     /* NOTE: Don't trust face->size->metrics.ascender and descender values. */
-    if (dc->pdcattr->iGraphicsMode == GM_ADVANCED)
+    if (pdcattr->iGraphicsMode == GM_ADVANCED)
     {
         pmxWorldToDevice = DC_pmxWorldToDevice(dc);
         FtSetCoordinateTransform(face, pmxWorldToDevice);
```
Comment (learn-more) [inline win32ss/gdi/ntgdi/freetype.c:6129]: Do these freetype functions require a lock?

Comment (learn-more) [reply win32ss/gdi/ntgdi/freetype.c:6129]: they do, and it's locked I see.

File: win32ss/gdi/ntgdi/freetype.c [hunk 6]
```diff
@@ -6100,246 +6152,61 @@ IntExtTextOutW(
...
-                Scale = pdcattr->mxWorldToDevice.efM11;
-                if (FLOATOBJ_Equal0(&Scale))
-                    FLOATOBJ_Set1(&Scale);
-
-                /* do the shift before multiplying to preserve precision */
-                FLOATOBJ_MulLong(&Scale, Dx[i<<DxShift] << 6);
-                TextLeft += FLOATOBJ_GetLong(&Scale);
-                DPRINT("New TextLeft2: %I64d\n", TextLeft);
-            }
+        IntEngBitBlt(SurfObj,
```
Comment (learn-more) [inline win32ss/gdi/ntgdi/freetype.c:6196]: I am unsure if calling functions like that is going to be a problem, while we hold the freetype lock.

File: win32ss/gdi/ntgdi/freetype.c (hunk 4 above)
Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c:5901]: OK. Done.

File: win32ss/gdi/ntgdi/freetype.c (hunk 3 above)
Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c:5938]: https://freetype.org/freetype2/docs/tutorial/step2.html

```txt
    /* increment pen position --                       */
    /* we don't have access to a slot structure,       */
    /* so we have to use advances from glyph structure */
    /* (which are in 16.16 fixed float format)         */
    pen.x += image.advance.x >> 10;
    pen.y += image.advance.y >> 10;
```
`advance.x` is in 16.16 fixed float format.
So shifting 16 bits right becomes 26.6 fixed float format.
It's precise enough if the calcuration was not wrong.

File: win32ss/gdi/ntgdi/freetype.c (hunk 6 above)
Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c:6198]: As far as I know, they are independent of each other.

Comment (learn-more) [reply win32ss/gdi/ntgdi/freetype.c]: Indeed, but it is an 'external' function.
And if that function calls back into font stuff again, that might cause a problem.

**Comment by katahiromz** [comment]
Now I have disabled inline.

Test KVM BEFORE: 79b0fce https://build.reactos.org/#/builders/6/builds/32468 id=84948
Test KVM AFTER: 
https://reactos.org/testman/compare.php?ids=84948,???

Test VBox BEFORE: 79b0fce https://build.reactos.org/#/builders/3/builds/19945 id=84949
Test VBox AFTER: 
https://reactos.org/testman/compare.php?ids=84949,???

**Comment by katahiromz** [comment]
Test KVM BEFORE: 79b0fce https://build.reactos.org/#/builders/6/builds/32468 id=84948
Test KVM AFTER: d24cf47 id=???
https://reactos.org/testman/compare.php?ids=84948,???

Test VBox BEFORE: 79b0fce https://build.reactos.org/#/builders/3/builds/19945 id=84949
Test VBox AFTER: d24cf47 id=???
https://reactos.org/testman/compare.php?ids=84949,???

**Comment by katahiromz** [comment]
Test KVM BEFORE: 79b0fce https://build.reactos.org/#/builders/6/builds/32468 id=84948
Test KVM AFTER: b9e4e31 id=???
https://reactos.org/testman/compare.php?ids=84948,???

Test VBox BEFORE: 79b0fce https://build.reactos.org/#/builders/3/builds/19945 id=84949
Test VBox AFTER: b9e4e31 id=???
https://reactos.org/testman/compare.php?ids=84949,???

**Comment by katahiromz** [comment]
Double check.

Test KVM BEFORE: 79b0fce https://build.reactos.org/#/builders/6/builds/32468 id=84948
Test KVM AFTER: eed3c22 id=???
https://reactos.org/testman/compare.php?ids=84948,???

Test VBox BEFORE: 79b0fce https://build.reactos.org/#/builders/3/builds/19945 id=84949
Test VBox AFTER: eed3c22 id=???
https://reactos.org/testman/compare.php?ids=84949,???

**Comment by katahiromz** [comment]
Total Commander:
Total

**Comment by katahiromz** [comment]
ftGdiGetTextWidth enabled.

Test KVM BEFORE: 79b0fce https://build.reactos.org/#/builders/6/builds/32468 id=84948
Test KVM AFTER: 2c89c91 id=???
https://reactos.org/testman/compare.php?ids=84948,???

Test VBox BEFORE: 79b0fce https://build.reactos.org/#/builders/3/builds/19945 id=84949
Test VBox AFTER: 2c89c91 id=???
https://reactos.org/testman/compare.php?ids=84949,???

**Comment by katahiromz** [comment]
VirtualBox_ReactOS_24_11_2022_11_40_32
RawTherapee is quickly working...

**Comment by HBelusca** [comment]
Why this has been merged and the flag "review pending" removed since nobody reviewed the PR yet?

File: win32ss/gdi/ntgdi/freetype.c [hunk 7]
```diff
@@ -6601,12 +6380,63 @@ IntExtTextOutW(
...
+                          (TextLeft64 + 32) >> 6,
+                          TextTop + yoff - underline_position + i,
+                          NULL,
+                          ROP2_TO_MIX(R2_COPYPEN));
+            }
+        }
+
+        if (plf->lfStrikeOut) /* Draw strike-out */
+        {
+            for (i = -thickness / 2; i < -thickness / 2 + thickness; ++i)
```
Comment (HBelusca) [inline win32ss/gdi/ntgdi/freetype.c:6427]: `-thickness / 2 + thickness` ?  Not just `+thickness / 2` ?

**Comment by HBelusca** [comment]
At some point it would be of interest to extract out that freetype support and implement external font drivers, so that it could also be fully tested on Windows. This would allow to know whether all those fixes etc. are indeed correct.

File: win32ss/gdi/ntgdi/freetype.c (hunk 7 above)
Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c:6427]: In C/C++, `(1 / 2)` is zero.

**Pull request merged**